Plain data objects

goto-bus-stop opened the pull request:
Implements these 'plain object' operations:
- creating an empty object
- getting and setting properties, named properties, index properties
- listing properties

And the required tag checks to make that work, and the required array methods to make `get_own_property_names` usable.

Most of it is quite straightforward, just calling napi functions. The remaining changes are about passing `Env` values to functions that need it under n-api, but did not need it under nan. The test for the object code uses downcasting, and the downcasting and typecheck code did not previously receive a reference to the `Env`. The `js_array.len()` function also did not receive a reference to the `Env`, but needs one to eventually call the napi function.

For the `js_array.len()` function, which is used by other methods on the impl, I added a private `len_inner()` function with a consistent API to save some feature flag checks. It could be removed if the nan implementation is removed later.

dherman commented on crates/neon-runtime/src/napi/tag.rs:
```diff
@@ -1,23 +1,40 @@
+}
 
-pub unsafe extern "C" fn is_function(_val: Local) -> bool { unimplemented!() }
+pub unsafe extern "C" fn is_object(_env: Env, _val: Local) -> bool { unimplemented!() }
```
Any reason you didn't implement this one yet? Not a big deal, we can take it incrementally, just curious.

goto-bus-stop replied:
Since it wasn't the focus of the PR, I only implemented the ones that were required to make the tests work. #507 includes a few more already

goto-bus-stop commented on crates/neon-runtime/src/napi/array.rs:
```diff
@@ -1,5 +1,14 @@
+/// Gets the length of a `napi_value` containing a JavaScript Array.
+pub unsafe extern "C" fn len(env: Env, array: Local) -> u32 {
+    let mut len = 0;
+    assert_eq!(napi::napi_get_array_length(env, array, &mut len as *mut _), napi::napi_status::napi_ok);
```
A convention I've seen a few times is putting `# Panics` sections in doc comments, so I've added both the non-`Array` and the pending exception case there. That does mean that a case that users actively need to think about (`isArray`) and a case that's quite rare are given roughly the same prominence—do you think it's clear enough like this?

dherman replied:
That's perfect, love it 👍

goto-bus-stop commented:
Test failure is because Node.js 10 appears to return a Number in a mixed `get_own_property_names` result (`[0, "a", "whatever"]` in the test case), while Node.js 12 casts them all to strings (`["0", "a", "whatever"]`). https://github.com/nodejs/node/pull/27524

`Object.getOwnPropertyNames()` returns only strings and the n-api documentation also claims it returns an array of strings, so I'll add some code to convert numeric indices to strings in those older node versions.

goto-bus-stop commented:
There was another complication: `get_property_names` includes inherited names, so I also added some `has_own_property` checks. That required one of two things:
- Removing elements from the array
- Creating a second array that only includes own properties

I tried the first approach using `napi_delete_property`, but that just resulted in holey arrays; I think I'd have to call the `.splice()` method from Rust to make that work and that is kind of difficult (and might not be very fast…).
So, I opted for the second.

I did work on a test for it too, but it requires a way to call into Rust functions from JS, which isn't yet implemented in this PR. I rebased https://github.com/neon-bindings/neon/pull/507 and added a test there.

One final thing to check: does `get_property_names` include Symbol properties? Its name suggests no, but let's find out :joy:

dherman commented:
@goto-bus-stop I'll review more closely in a bit, thanks for updating!

Technically, the semantics of `getOwnPropertyNames` isn't the same as doing a full enumeration and then filtering out anything that isn't an own property. For example, a proxy could behave very differently.

But it looks like N-API simply doesn't expose the `getOwnPropertyNames` operation, so this might be the closest we can come to matching the right behavior for now. I've [asked some Node and N-API leaders](https://twitter.com/littlecalculist/status/1248378005161775105?s=20) how we might propose a new N-API function to do what we need.

dherman commented:
Per the replies on Twitter, I've [filed an N-API issue](https://github.com/nodejs/node-addon-api/issues/698) proposing an API.

dherman commented:
@goto-bus-stop Oh, wait! Twitter [came to the rescue](https://twitter.com/gabrielschulhof/status/1248388727635771393) and explained that [`napi_get_all_property_names()`](https://nodejs.org/api/n-api.html#n_api_napi_get_all_property_names) is the API we need here. You can pass a `key_mode` value of `napi_key_own_only` and a `key_filter` of `napi_key_skip_symbols`. That will have exactly the behavior of `Object.getOwnPropertyNames()`.

goto-bus-stop commented:
Riiight Proxies…I really should've read the spec instead of guessing!

I wanted to use `napi_get_all_property_names` first, but the documentation said it's only supported in Node.js 13+, which was a dealbreaker. However, looking at Node.js commit history now, it looks like it's actually been backported to 10.x too, so that makes things easier :sweat_smile: 
Just needs a nodejs-sys update to add the newer APIs.

goto-bus-stop commented:
Oh lol—the backport was released in `10.20.0` only 2 days ago, which is why I didn't know about it yet :)

goto-bus-stop commented:
`napi_get_all_property_names` is not yet in Node.js 12.x but may be backported in the next minor in April per https://github.com/nodejs/node/pull/30006#issuecomment-596710840

goto-bus-stop commented:
I guess tests won't pass on Node 12 until a new minor is released. Should we in the mean time just rely on my buggy implementation, and file a follow-up issue to upgrade as soon as a Node.js 12.x with the backport is available?

dherman commented:
@goto-bus-stop Thanks for all this detective work! I think your workaround is totally fine for now until the new API is backported to 12. Do you mind also filing a followup issue so we can replace the workaround once 12 is ready?

goto-bus-stop commented:
filed follow up in https://github.com/neon-bindings/neon/issues/511
